### PR TITLE
Created starting delivery and plugin system

### LINKED ISSUE #13: sensor delivery package
should have the count of people, sensor id, and should be extendable for other pieces of data

### PR DESCRIPTION
**Pull request by mooncollin**
Created primitive plugin system for the seer platform. Plugins are created by putting python files in the plugins folder and it will automatically pick up any classes that inherit from the data_collector_plugin class.

Examples are in the test.py plugin. Currently, the delivery system is just printing out to the console. It will eventually be communicating JSON to a server.

### CODE REVIEW

File: seers/plugins/test.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
+from seer import data_collector_plugin
```
Comment (pgayee694) [inline seers/plugins/test.py]: this should be moved into it's own directory like `tests/`

File: seers/plugins/__init__.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+PYTHON_EXTENSION = '.py'
+IGNORED_FILES = ['__init__.py']
+
+modules = glob.glob(os.path.join(os.path.dirname(__file__), '*' + PYTHON_EXTENSION))
+
+__all__ = [os.path.basename(f)[:-len(PYTHON_EXTENSION)] for f in modules if os.path.isfile(f) and os.path.basename(f) not in IGNORED_FILES]
+from . import *
+
+def gather_plugins():
+	plugins = []
```
Comment (pgayee694) [inline seers/plugins/__init__.py]: documentation

File: seers/seer.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
+class data_collector_plugin(object):
+	def __init__(self):
```
Comment (pgayee694) [inline seers/seer.py]: documentation on the object and the rest of the functions/object declarations in this file

File: seers/seer.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+	def init(self):
+		pass
+
+	def shutdown(self):
+		pass
+
+	def collect(self):
+		return {}
+
+import plugins
```
Comment (pgayee694) [inline seers/seer.py]: imports should be at the top of the file

File: seers/seer.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+			try:
+				message_queue.put(data, True, .2)
+				break
+			except queue.Full:
+				pass
+		time.sleep(time_step / 1000)
+
+	plug.shutdown()
+
+if __name__ == '__main__':
```
Comment (pgayee694) [inline seers/seer.py]: move this to it's own file

File: seers/plugins/test.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+		if self.opened:
+			data = self.video.read()
+
+		return {"camera": self.opened}
+
+	def shutdown(self):
+		if self.opened:
+			self.video.release()
+
+# class camera_collector2(data_collector_plugin):
```
Comment (pgayee694) [inline seers/plugins/test.py]: why is this commented out?

**Review by pgayee694 — CHANGES_REQUESTED** [review]

File: seers/plugins/__init__.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+import sys
+import inspect
+import seer_plugin
+
+PYTHON_EXTENSION 	= '.py'
+IGNORED_FILES 		= ['__init__.py']
+
+modules = glob.glob(os.path.join(os.path.dirname(__file__), '*' + PYTHON_EXTENSION))
+
+__all__ = [os.path.basename(f)[:-len(PYTHON_EXTENSION)] for f in modules if os.path.isfile(f) and os.path.basename(f) not in IGNORED_FILES]
```
Comment (pgayee694) [inline seers/plugins/__init__.py]: don't search for files, have the entry point instantiate all the seers you want to collect and reference/manipulate them that way; then we can just maintain that along with enums

File: seers/plugins/tests/test.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
+import seer_plugin
```
Comment (pgayee694) [inline seers/plugins/tests/test.py]: might want to reconsider naming this file to something like "test_plugin" or something to differentiate it from unit tests and then have unit tests that use it to test the functionality

File: seers/seer.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+import sys
+import logging
+import requests
+
+logging.basicConfig(level=seer_config.logging_level, format='[%(levelname)s] %(asctime)s (%(module)s): %(message)s')
+logging.getLogger('requests').setLevel(logging.WARNING)
+logging.getLogger('urllib3').setLevel(logging.WARNING)
+
+http_session	= requests.Session()
+timeout			= None if seer_config.data_collection_timeout is None else seer_config.data_collection_timeout / 1000
```
Comment (pgayee694) [inline seers/seer.py:30]: /nit wierd spacing?

File: seers/seer_config.py [hunk 10]
```diff
@@ -0,0 +1,19 @@
+import configparser
+import logging
+
+configuration = configparser.ConfigParser()
+configuration.read('seer_config.ini')
+
+general_ini					= 'general'
```
Comment (pgayee694) [inline seers/seer_config.py:7]: /nit wierd spacing again?

File: seers/plugins/tests/test.py [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+import random
+import cv2
+import logging
+
+class test_collector(object):
```
Comment (pgayee694) [inline seers/plugins/tests/test.py]: classes should be renamed in pep8 convention, i.e. TestCollector

File: seers/seer_config.py [hunk 12]
```diff
@@ -0,0 +1,19 @@
+import configparser
```
Comment (pgayee694) [inline seers/seer_config.py:1]: Should this be a class?

File: seers/seer_plugin.py [hunk 13]
```diff
@@ -0,0 +1,77 @@
+import seer_config
+
+import queue
+import time
+import logging
+
+class data_collector_plugin(object):
```
Comment (pgayee694) [inline seers/seer_plugin.py]: /nit Consider naming this something like "AbstractDataCollectorPlugin" or "DataCollectorPluginBase" to make sure it's easily recognizable as the base class

**Review by pgayee694 — CHANGES_REQUESTED** [review]

File: seers/seer_plugin.py [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+
+	def collect(self):
+		"""
+		Called every time step to collect data. Must
+		return a map consisting of named data. This data
+		will eventually be delivered to the Delphi system.
+		"""
+		return {}
+
+def plugin_collection(plug_class, message_queue, stop_token):
```
Comment (svnphlps) [inline seers/seer_plugin.py:45]: Maybe rename this function to 'collect_data'.

File: seers/seer_config.py (hunk 10 above)
Comment (mooncollin) [reply seers/seer_config.py:7]: Git does that. It's not the source

**Review by pgayee694 — APPROVED** [review]

**Review by svnphlps — APPROVED** [review]

**Pull request merged**